pkg/atomicwriter/atomic_writer.go

```diff
@@ -0,0 +1,108 @@
+package atomicwriter
+
+import (
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/cyberark/conjur-authn-k8s-client/pkg/log"
+	"github.com/cyberark/secrets-provider-for-k8s/pkg/log/messages"
+)
+
+const deleteTempFileRetryInterval = time.Second * 5
+const deleteTempFileMaxRetries = 5
+
+// This package provides a simple atomic file writer which implements the
+// io.WriteCloser interface. This allows us to use AtomicWriter the way we
+// would use any other Writer, such as a Buffer. Additonally, this struct
+// takes the file path during construction, so the code which calls
+// `Write()` doesn't need to be concerned with the destination, just like
+// any other writer.
+type AtomicWriter struct {
+	path        string
+	permissions os.FileMode
+	tempFile    *os.File
+}
+
+func NewAtomicWriter(path string, permissions os.FileMode) (*AtomicWriter, error) {
+	dir, file := filepath.Split(path)
+	if dir == "" {
+		dir = "."
+	}
+
+	f, err := ioutil.TempFile(dir, file)
+	if err != nil {
+		log.Error(messages.CSPFK055E, path)
+		return nil, err
+	}
+
+	return &AtomicWriter{
+		path:        path,
+		tempFile:    f,
+		permissions: permissions,
+	}, nil
+}
+
+func (w *AtomicWriter) Write(content []byte) (n int, err error) {
+	// Write to the temporary file
+	return w.tempFile.Write(content)
+}
+
+func (w *AtomicWriter) Close() error {
+	defer func() {
+		go w.Cleanup()
+	}()
```

Comment:
1. Why does this need retry logic ?
2. Deferring a call to a go routine feels like trouble

Reply:
1. See [this thread](https://github.com/cyberark/secrets-provider-for-k8s/pull/428#discussion_r794975481)
2. Why? Is there a preferred way to do this? My assumption is that a regular defer() function will block execution, which is a problem if we're doing timed retries.

Reply:
1. I'm not sure about the appropriateness of retry here. I think this is an instance of an application assuming the reliability responsibilities of something like the OS/file system. I think we ought to be able to safely assume reliability here. I understand that deletion might fail, but I think a single attempt + errors should be enough. An alternative might be to try and delete and if it doesn't work Truncate the file to 0, which is highly likely to work if done after successful writes. 
2. It's sort of a code smell, at least in my eyes. Defer creating Go routines that are not tracked by anything... An alternative, if we really decide to keep an eye on those tmp file and make sure they are cleaned up might be to a have some sort of temp file manager that keeps track of a collection (maybe a queue) of temp files and in the background in a single goroutine shared between all the files can carry out the regular task of cleaning up the temp files.

Reply:
@andytinkham What do you think?

Reply:
It feels like truncate to 0 solves the problem of possibly leaking secrets which is my main concern. If that's easier, I'm ok with it.  For #2, I don't think I have any specific thoughts - seems OK.

Reply:
OK. So instead of the whole retry logic, we just
1. Attempt to delete the file
2. If that fails, attempt to truncate it (and maybe log it)
3. If that fails, log the error and exit
Is that right?